File: app/src/main/java/de/karzek/diettracker/presentation/main/diary/meal/adapter/diaryEntryList/viewHolder/DiaryEntryViewHolder.java
package de.karzek.diettracker.presentation.main.diary.meal.adapter.diaryEntryList.viewHolder;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.TextView;

import com.daimajia.swipe.SwipeLayout;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import de.karzek.diettracker.R;
import de.karzek.diettracker.data.cache.model.GroceryEntity;
import de.karzek.diettracker.presentation.model.DiaryEntryDisplayModel;
import de.karzek.diettracker.presentation.util.StringUtils;

/**
 * Created by MarjanaKarzek on 30.05.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 30.05.2018
 */
public class DiaryEntryViewHolder extends RecyclerView.ViewHolder {

    @BindView(R.id.grocery_summary) TextView grocerySummary;
    @BindView(R.id.swipe_layout) SwipeLayout swipeLayout;
    @BindView(R.id.swipe_option_move) ImageButton moveOption;

    private final OnDiaryEntryItemClickedListener onItemClickedListener;
    private final OnDeleteDiaryEntryItemListener onDeleteDiaryEntryItemListener;
    private final OnMoveDiaryEntryItemListener onMoveDiaryEntryItemListener;
    private final OnEditDiaryEntryItemListener onEditDiaryEntryItemListener;

    public DiaryEntryViewHolder(ViewGroup viewGroup,
                                OnDiaryEntryItemClickedListener onItemClickedListener,
                                OnDeleteDiaryEntryItemListener onDeleteDiaryEntryItemListener,
                                OnMoveDiaryEntryItemListener onMoveDiaryEntryItemListener,
                                OnEditDiaryEntryItemListener onEditDiaryEntryItemListener) {
        super(LayoutInflater.from(viewGroup.getContext())
                .inflate(R.layout.item_diary_grocery, viewGroup, false));
        ButterKnife.bind(this, itemView);
        this.onItemClickedListener = onItemClickedListener;
        this.onDeleteDiaryEntryItemListener = onDeleteDiaryEntryItemListener;
        this.onMoveDiaryEntryItemListener = onMoveDiaryEntryItemListener;
        this.onEditDiaryEntryItemListener = onEditDiaryEntryItemListener;
    }

    public void bind(DiaryEntryDisplayModel diaryEntry) {
        grocerySummary.setText(formatGrocerySummary(diaryEntry));
        itemView.setTag(diaryEntry.getId());

        if(diaryEntry.getGrocery().getType() == GroceryEntity.GroceryEntityType.TYPE_DRINK)
            moveOption.setVisibility(View.GONE);
    }

    private String formatGrocerySummary(DiaryEntryDisplayModel diaryEntry){
        String amount = StringUtils.formatFloat(diaryEntry.getAmount());
        return "" + amount + diaryEntry.getUnit().getName() + " " + diaryEntry.getGrocery().getName();
    }

    @OnClick(R.id.grocery_summary) public void onItemClicked() {
        onItemClickedListener.onDiaryEntryClicked((int) itemView.getTag());
    }

    @OnClick(R.id.swipe_option_delete) public void onItemDeleteClicked(){
        swipeLayout.close(false);
        onDeleteDiaryEntryItemListener.onDiaryEntryDeleteClicked((int) itemView.getTag());
    }

    @OnClick(R.id.swipe_option_move) public void onItemMoveClicked(){
        swipeLayout.close();
        onMoveDiaryEntryItemListener.onDiaryEntryMoveClicked((int) itemView.getTag());
    }

    @OnClick(R.id.swipe_option_edit) public void onItemEditClicked(){
        onEditDiaryEntryItemListener.onDiaryEntryEditClicked((int) itemView.getTag());
    }

    public interface OnDiaryEntryItemClickedListener {
        void onDiaryEntryClicked(int id);
    }

    public interface OnDeleteDiaryEntryItemListener {
        void onDiaryEntryDeleteClicked(int id);
    }

    public interface OnMoveDiaryEntryItemListener {
        void onDiaryEntryMoveClicked(int id);
    }

    public interface OnEditDiaryEntryItemListener {
        void onDiaryEntryEditClicked(int id);
    }

}

File: app/src/main/java/de/karzek/diettracker/presentation/main/diary/meal/adapter/diaryEntryList/viewHolder/DiaryEntryViewHolder.kt
package de.karzek.diettracker.presentation.main.diary.meal.adapter.diaryEntryList.viewHolder

import android.view.View
import androidx.recyclerview.widget.RecyclerView
import de.karzek.diettracker.data.cache.model.GroceryEntity
import de.karzek.diettracker.presentation.model.DiaryEntryDisplayModel
import de.karzek.diettracker.presentation.util.StringUtils
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.item_diary_grocery.*

/**
 * Created by MarjanaKarzek on 30.05.2018.
 *
 * @author Marjana Karzek
 * @version 1.0
 * @date 30.05.2018
 */
class DiaryEntryViewHolder(override val containerView: View,
                           private val onItemClickedListener: OnDiaryEntryItemClickedListener,
                           private val onDeleteDiaryEntryItemListener: OnDeleteDiaryEntryItemListener,
                           private val onMoveDiaryEntryItemListener: OnMoveDiaryEntryItemListener,
                           private val onEditDiaryEntryItemListener: OnEditDiaryEntryItemListener) : RecyclerView.ViewHolder(containerView), LayoutContainer {

    fun bind(diaryEntry: DiaryEntryDisplayModel) {
        grocery_summary.text = formatGrocerySummary(diaryEntry)
        itemView.tag = diaryEntry.id

        if (diaryEntry.grocery.type == GroceryEntity.GroceryEntityType.TYPE_DRINK)
            swipe_option_move.visibility = View.GONE

        setupClickListeners()
    }

    private fun formatGrocerySummary(diaryEntry: DiaryEntryDisplayModel): String {
        val amount = StringUtils.formatFloat(diaryEntry.amount)
        return "" + amount + diaryEntry.unit.name + " " + diaryEntry.grocery.name
    }

    private fun setupClickListeners() {
        grocery_summary.setOnClickListener { onItemClickedListener.onDiaryEntryClicked(itemView.tag as Int) }
        swipe_option_delete.setOnClickListener {
            swipe_layout.close(false)
            onDeleteDiaryEntryItemListener.onDiaryEntryDeleteClicked(itemView.tag as Int)
        }
        swipe_option_move.setOnClickListener {
            swipe_layout.close()
            onMoveDiaryEntryItemListener.onDiaryEntryMoveClicked(itemView.tag as Int)
        }
        swipe_option_edit.setOnClickListener { onEditDiaryEntryItemListener.onDiaryEntryEditClicked(itemView.tag as Int) }
    }

    interface OnDiaryEntryItemClickedListener {
        fun onDiaryEntryClicked(id: Int)
    }

    interface OnDeleteDiaryEntryItemListener {
        fun onDiaryEntryDeleteClicked(id: Int)
    }

    interface OnMoveDiaryEntryItemListener {
        fun onDiaryEntryMoveClicked(id: Int)
    }

    interface OnEditDiaryEntryItemListener {
        fun onDiaryEntryEditClicked(id: Int)
    }

}
